backend/src/lib.rs

```diff
@@ -1,104 +1,186 @@
+#[cfg(feature = "halo2")]
+mod halo2_structs;
+mod pilcom_cli;
+
 use ast::analyzed::Analyzed;
-use number::FieldElement;
-use std::io;
+use number::{DegreeType, FieldElement};
+use std::{io, marker::PhantomData, path::Path};
 use strum::{Display, EnumString, EnumVariantNames};
 
 #[derive(Clone, EnumString, EnumVariantNames, Display)]
-pub enum Backend {
+pub enum BackendType {
     #[cfg(feature = "halo2")]
     #[strum(serialize = "halo2")]
     Halo2,
     #[cfg(feature = "halo2")]
-    #[strum(serialize = "halo2-aggr")]
-    Halo2Aggr,
-    #[cfg(feature = "halo2")]
     #[strum(serialize = "halo2-mock")]
     Halo2Mock,
-    // At the moment, this enum is empty without halo2, but it is always built
-    // as part of the infrastructure to eventually support other backends.
+    #[strum(serialize = "pilcom-cli")]
+    PilcomCli,
 }
 
-/// Create a proof for a given PIL, fixed column values and witness column values
-/// using the chosen backend.
+impl BackendType {
+    pub fn build<T: FieldElement>(&self) -> &'static dyn BackendFactory<T> {
+        #[cfg(feature = "halo2")]
+        const HALO2_FACTORY: WithSetupFactory<halo2::Halo2Prover> = WithSetupFactory(PhantomData);
+        #[cfg(feature = "halo2")]
+        const HALO2_MOCK_FACTORY: WithoutSetupFactory<halo2_structs::Halo2Mock> =
+            WithoutSetupFactory(PhantomData);
+        const PILCOM_CLI_FACTORY: WithoutSetupFactory<pilcom_cli::PilcomCli> =
+            WithoutSetupFactory(PhantomData);
 
-pub type Proof = Vec<u8>;
-pub type Params = Vec<u8>;
+        match self {
+            #[cfg(feature = "halo2")]
+            BackendType::Halo2 => &HALO2_FACTORY,
+            #[cfg(feature = "halo2")]
+            BackendType::Halo2Mock => &HALO2_MOCK_FACTORY,
+            BackendType::PilcomCli => &PILCOM_CLI_FACTORY,
+        }
+    }
+}
 
-pub trait ProverWithParams {
-    fn prove<T: FieldElement, R: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof>;
+/// Factory for backends without setup.
+struct WithoutSetupFactory<B>(PhantomData<B>);
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params;
-}
+/// Factory implementation for backends without setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F> + 'static> BackendFactory<F>
+    for WithoutSetupFactory<B>
+{
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
```

Comment:
Why not just 
```suggestion
    fn new(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
```
?
Usually we refer to the setup artifact as "params", I think just `new` is fine here